### PR TITLE
NW - Terry Calderbank - JavaScript-Core-1-Coursework-Week3

### PR DESCRIPTION
**Pull request by Texx45**
**Volunteers: Are you marking this coursework?** _You can find a guide on how to mark this coursework in `HOW_TO_MARK.md` in the root of this repository_

# Your Details

- Your Name:
- Your City:
- Your Slack Name:

# Homework Details

- Module:
- Week:

# Notes

- What did you find easy?

- What did you find hard?

- What do you still not understand?

- Any other notes?

-----
[View rendered 1-exercises/A-array-find/README.md](https://github.com/Texx45/JavaScript-Core-1-Coursework-Week3/blob/main/1-exercises/A-array-find/README.md)
[View rendered 1-exercises/H-array-methods-2/README.md](https://github.com/Texx45/JavaScript-Core-1-Coursework-Week3/blob/main/1-exercises/H-array-methods-2/README.md)

### CODE REVIEW

File: 2-mandatory/6-journey-planner.js [hunk 1]
```diff
@@ -126,15 +126,23 @@ function getLocationName(locationName) {
...
-  // Implement the function body
-  locations
-    .filter((location) => {
-      return isAccessibleByTransportMode(location, transportMode);
-    })
-    .map((location) => getLocationName(location));
+  return locations
+    .filter((location) => isAccessibleByTransportMode(location, transportMode))
+    .map((possibleLocations) => getLocationName(possibleLocations));
 }
```
Comment (gulnihaln) [inline 2-mandatory/6-journey-planner.js:142]: I can see here a brief and clean code Terry well done!

**Comment by gulnihaln** [comment]
Great job Terry ! nicely done and I know how you worked on some specific exercises this week, well done.

File: 2-mandatory/6-journey-planner.js (hunk 1 above)
Comment (Texx45) [reply 2-mandatory/6-journey-planner.js:142]: Thanks Gulnihal! it was hard but I wont take credit as we did it as a group with Michael.

**Comment by SteveLeicester** [comment]
Thanks Terry.

All the exercises look and work fine.

In mandatory challenges - 
the 1-create-function.js file has two `sortArray` functions and no remove. So perhaps you copied began to modify and then ran out of time. But actually the second version is correct unlike the first

`myArray.sort()`

and

`[...myArray].sort()`

The spread operator is the way to do it - i.e. get a copy (or actually slice() - with no arguments means slice from start to end of the array). Then sort the copy.  The first version modifies the original array.  May be that is what is needed but often a source of bugs (things getting modified unexpectedly - see lower down).


So I think there are several things with arrays resulting in it being challenging.

1.  The number of operations or methods that you can do with them.
2.  Which ones modify (mutate) the original array and which ones create a new version leaving the original array intact.  This is usually very important. If you write some (library) code that someone else is going to use and they pass into your library of very useful functions some data (in the form of an array)they might be surprised and upset that their original data has changed.
3.  Hence - how to make a copy if you don't want the array to change

I think getting used to all these and what they do does take a little time.  A cheat sheet is always helpful but I really liked this pictorial view. It's a bit limited but really helps especially if your mind and learning work more visually.  I'll see if I can find any more.

[https://twitter.com/calebareeveso/status/1343619471597035520/photo/1](https://twitter.com/calebareeveso/status/1343619471597035520/photo/1)

Another thought was I had was around methods which

1.  Changes things in an array into something else (map)
2.  Insertions and deletions (eg filter, splice)
3.  Searching and Sorting (eg find, include)
4.  Collapses into a single value (reduce)

So trying to conceptualise around those areas.




One other thing here - Looking at format percentage - yours fails the test. Firstly cap at 100% which is straight forward

`element = element > 100 ? 100 : element;`

But actually I think the test doesn't match the instruction. So it says "The numbers must be rounded to 2 decimal places". And you did  `toFixed(2)` and `33` comes out as `"33.00%"`. But the test says it should be `"33%"` which is definitely not 33 to 2 decimal places.  So I'd take take that one definitely. 




On the journey planner. I know you worked with others - wondering if thinking around the different concepts would help in tackling this.  In any case you might surprise yourself if you tackle other similar tasks on how much more you know and what seemed difficult is perhaps not so any more.

Your password validator works - you look in the array and get the index of the current password and compare with the passed in index. If they don't match then there must be another instance of the same password. Given that `indexOf` always returns the index of the first instance even if there is more than one then it is bound to find there is a later duplicate password.  Its a great solution.  When I thought about it my initial thoughts were to remove the current password (so a copy of the array) then see if the current password (now removed) is included in this new array.  I've a feeling your solution would be more efficient so excellent (making copies means using more memory). 

Nothing really to add regarding the other challenges and I see you did the extra sorting challenge too. Although it may have taken you more time than you might have wished I think this is a great submission so well done.

Steve

**Comment by Texx45** [comment]
Hi Steve, thanks for the feedback. We have a quieter week as we are doing some PD stuff this week. I am taking the opportunity to go over array methods before we move onto the next topic and it is making a bit more sense so far.  I'm not sure what happened with Format Percentage I guess I didn't read it correctly but I was certain they had passed all the tests but obviously not! I think I need to pay more attention to detail. 

Regarding the check for the password duplication I would have done it the same way to be honest so I can't take credit for that. That was the part I was having an issue with and one of my peers Mya helped me with that.

Thanks once again for going over my homework I know it must be time consuming. I am learning so much even even I do find some parts incredibly difficult.

**Pull request closed without merging**